ott/core/initializers.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2022 The OTT Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Sinkhorn initializers."""
+from typing import Optional, Tuple
+
+import jax
+import jax.numpy as jnp
+
+from ott.core import linear_problems
+from ott.geometry.pointcloud import PointCloud
+
+
+def default_dual_a(
+    ot_problem: linear_problems.LinearProblem, lse_mode: bool
+) -> jnp.ndarray:
+  """Return dual potential vector, f.
+
+  Args:
+    ot_problem:
+    lse_mode: Return log potentials if true. Defaults to True.
+
+  Returns:
+    potential f, array of size n
+  """
+  a = ot_problem.a
+  init_dual_a = jnp.zeros_like(a) if lse_mode else jnp.ones_like(a)
+  return init_dual_a
+
+
+def default_dual_b(
+    ot_problem: linear_problems.LinearProblem, lse_mode: bool
+) -> jnp.ndarray:
+  """Return dual potential vector, g.
+
+  Args:
+    ot_problem:
+    lse_mode: Return log potentials if true. Defaults to True.
+
+  Returns:
+    potential g, array of size m
+  """
+  b = ot_problem.b
+  init_dual_b = jnp.zeros_like(b) if lse_mode else jnp.ones_like(b)
+  return init_dual_b
+
+
+def remove_weight_potential(
+    weights: jnp.ndarray, init_dual: jnp.ndarray, lse_mode: bool
+) -> Tuple[jnp.ndarray]:
+  """Cancel dual variables for zero weights.
+
+  Args:
+    weights: array of probability masses
+    init_dual: dual potential array
+    lse_mode: Return log potentials if true. Defaults to True.
+
+  Returns:
+    potential
+  """
+  return jnp.where(weights > 0, init_dual, -jnp.inf if lse_mode else 0.0)
+
+
+def remove_weight_potentials(
+    weights_a: jnp.ndarray, weights_b: jnp.ndarray, init_dual_a: jnp.ndarray,
+    init_dual_b: jnp.ndarray, lse_mode: bool
+) -> Tuple[jnp.ndarray]:
+  """Cancel dual variables for zero weights.
+
+  Args:
+    weights_a: array of probability masses, array of size n
+    weights_b: array of probability masses, array of size m
+    init_dual_a: potential f, array of size n
+    init_dual_b: potential g, array of size m
+    lse_mode: Return log potentials if true. Defaults to True.
+
+  Returns:
+    potentials (f,g)
+  """
+  init_dual_a = remove_weight_potential(weights_a, init_dual_a, lse_mode)
+  init_dual_b = remove_weight_potential(weights_b, init_dual_b, lse_mode)
+  return init_dual_a, init_dual_b
+
+
+class SinkhornInitializer:
+  """Initialzation for Sinkhorn potential f.
+
+  Args:
+    ot_problem: OT problem between discrete distributions of size n and m.
+    lse_mode: Return log potential. Defaults to True.
```

Comment:
In fact lse_mode outputs a potential, lse_mode False outputs exp(potential/epsilon)